book/src/confidence.md

```diff
@@ -0,0 +1,67 @@
+
+# Computation of confidence metric for banks >= root
+
+Builiding this hashmap leverages the computations already being made for building consensus. The `collect_vote_lockouts` function in `consensus.rs` builds a HashMap, where each entry is of the form `(b, s)` where `s` is a `StakeLockout` struct representing the amount of stake and lockout on a bank `b`.
+
+This computation is performed on a votable candidate bank `b` as follows.
+
+```
+   let output: HashMap<b, StakeLockout> = HashMap::new();
+   for vote_account in b.vote_accounts {
+       for v in vote_account.vote_stack {
+           for a in ancestors(v) {
+               f(*output.get_mut(a), vote_account, v);
+           }
+       }
+   }
+```
+
+where `f` is some accumulation function that modifies the `StakeLockout` entry for slot `a` with some data derivable from vote `v` and `vote_account` (stake, lockout, etc.). Note here that the `ancestors` here only includes slots >= the current root. 
+
+Now we can naturally augment the above computation to also build a `Confidence` hashmap for every bank `b` by: 
+1) Augmenting `StakeLockout` to include a `confidence` field that represents a `Confidence` hashmap
+2) Replacing `f` with `f'` such that the above computation also builds this `Confidence` hashmap for every bank `b`.
+
+We will proceed with the details of 2) as 1) is trivial.
+
+Before continuing, it is noteworthy that for some validator's vote account `a`, the number of local confirmations for that validator on slot `s` is `v.num_confirmations`, where `v` is the smallest vote in the stack of votes `a.votes` such that `v.slot >= s` (i.e. there is no need to look at any votes > v as the number of confirmations will be lower).
+
+Now more specifically, we augment the above computation to:
+
+```
+   let output: HashMap<b, StakeLockout> = HashMap::new();
+   for vote_account in b.vote_accounts {
+       // vote stack is sorted from oldest vote to newest vote
+       for (v1, v2) in vote_account.vote_stack.windows(2) {
+           for a in ancestors(v1).difference(ancestors(v2)) {
+               f'(*output.get_mut(a), vote_account, v);
+           }
+       }
+   }
+```
+
+where `f'` is defined as:
+```
+    fn f`(some_ancestor: &mut StakeLockout, vote_account: VoteAccount, v: Vote, total_stake: u64) {
+        f(some_ancestor, vote_account, v);
+        *some_ancestor.confidence.entry(v.num_confirmations) += vote_account.stake / total_stake;
+    }
+```
+
+# Computation of confidence metric for banks < root
+In the above section, the `output` HashMap computed will only include entries for slots >= root b/c `ancestors` only contain entries >= root. This is a problem if clients want to call `get_confidence(s: Signature) -> Confidence` with a signature from a bank older than the root.
+
+TODO: Figure out what we want to do here
```

Comment:
All opinions welcome here :D

Reply:
confidence for slots < root is not exactly confidence of root?

Reply:
The aim is to provide a confidence metric that reflects cluster-wide finality on a slot. 

Root is a concept local to a node, and because the root is constantly moving, the confidence of the current root should be <= confidence of slots that come before the root, but that's not very precise.

For instance, let's say I'm interested in the cluster-wide confidence levels in block 5 and my root is moving 6, 7, 8, 9... 

If I just respond with the confidence in the current root, then the client will see fluctuating levels of confidence in block 5 corresponding to the confidence levels of blocks 6, 7, 8, 9.

Reply:
can we assume that 5 is a parent of 6, can we verify that 5 is a parent of 6?  we're returning information from vote accounts in Bank M, right?

seems to me that you don't need to limit the returned vector to MAX_LOCKOUT, the limiting factor is really the size of the signature cache and an ability to traverse ancestors past the roots implied by the vote accounts...

Reply:
Yes, you can verify 5 is a parent of 6. We limit the vector to MAX_LOCKOUT b/c the only valid lockouts are 0..MAX_LOCKOUT. You're right that the limiting factor is essentially the number of ancestors we want to perform processing for in `collect_vote_lockouts`. Right now b/c we're only processiinig up to our root (we filter out processing on anything < root), thats usually ~32 ancestors. If we increase to signature cache number of roots, that's 120 ancestors.